internal/command/meta_backend.go

```diff
@@ -97,14 +97,6 @@ func (m *Meta) Backend(opts *BackendOpts) (backend.Enhanced, tfdiags.Diagnostics
 		b, backendDiags = m.backendFromConfig(opts)
 		diags = diags.Append(backendDiags)
 
-		if opts.Init && b != nil && !diags.HasErrors() {
-			// Its possible that the currently selected workspace doesn't exist, so
-			// we call selectWorkspace to ensure an existing workspace is selected.
-			if err := m.selectWorkspace(b); err != nil {
-				diags = diags.Append(err)
-			}
-		}
```

Comment:
Are we confident that no other callers are relying on this behavior, being removed from this function? This feels like something that could be very quietly relied upon as a side effect.

Reply:
If I'm reading the code correctly, I don't think it has been removed—it moved into the `backendFromConfig` method, called a couple of lines up, with a slightly different conditional.

Reply:
@chrisarcand I am glad you ask! As I was making these changes, I try taking a close look into which of the return statements for `func (m *Meta) backendFromConfig(opts *BackendOpts) (backend.Backend, tfdiags.Diagnostics)` would make us reach this line of the code `if err := m.selectWorkspace(b); err != nil `.

So far I can see that four return statements will make us hit that conditional. These are those four statements:

https://github.com/hashicorp/terraform/blob/4ff0943e74f9f0fd02a400a8f43a3b60de7d8374/internal/command/meta_backend.go#L635

https://github.com/hashicorp/terraform/blob/4ff0943e74f9f0fd02a400a8f43a3b60de7d8374/internal/command/meta_backend.go#L710

https://github.com/hashicorp/terraform/blob/4ff0943e74f9f0fd02a400a8f43a3b60de7d8374/internal/command/meta_backend.go#L646

https://github.com/hashicorp/terraform/blob/4ff0943e74f9f0fd02a400a8f43a3b60de7d8374/internal/command/meta_backend.go#L669

And this is because all those 4 return statements can potentially return a backend.Backend that is not nil and where the user has run `init`, making this line asses to true:
https://github.com/hashicorp/terraform/blob/4ff0943e74f9f0fd02a400a8f43a3b60de7d8374/internal/command/meta_backend.go#L100

Considering those cases, I have called `selectWorkspace` within their path. Do you think there are other callers that I could be missing that are not those four paths? I'm glad you are taking a look at this, I am scared I could be missing something important!